src/webui/webapplication.cpp

```diff
@@ -915,76 +841,102 @@ void WebApplication::action_command_addCategory()
     BitTorrent::Session::instance()->addCategory(category);
 }
 
-void WebApplication::action_command_removeCategories()
+void WebApplication::removeCategories()
 {
-    CHECK_URI(0);
     CHECK_PARAMETERS("categories");
 
     QStringList categories = request().posts["categories"].split('\n');
     foreach (const QString &category, categories)
         BitTorrent::Session::instance()->removeCategory(category);
 }
 
-void WebApplication::action_command_getSavePath()
+void WebApplication::getDefaultSavePath()
 {
-    CHECK_URI(0);
     print(BitTorrent::Session::instance()->defaultSavePath());
 }
 
-bool WebApplication::isPublicScope()
+bool WebApplication::isPublicAPI(const QString &methodName) const
 {
-    return (scope_ == DEFAULT_SCOPE || scope_ == VERSION_INFO);
+    return m_publicAPIs.contains(methodName);
 }
 
 void WebApplication::doProcessRequest()
 {
-    scope_ = DEFAULT_SCOPE;
-    action_ = DEFAULT_ACTION;
-
-    parsePath();
+    if (request().path.startsWith(PATH_PREFIX_API)) {
+        const QString methodName {request().path.mid(PATH_PREFIX_API.size())};
+        if (!isPublicAPI(methodName) && !sessionActive()) {
+            status(403, "Forbidden");
+            return;
+        }
 
-    if (args_.contains(".") || args_.contains("..")) {
-        qDebug() << Q_FUNC_INFO << "Invalid path:" << request().path;
-        status(404, "Not Found");
-        return;
+        if (!QMetaObject::invokeMethod(this, methodName.toLatin1().constData())) {
```

Comment:
is `QMetaObject::invokeMethod()` possible to call to class methods? e.g. `doProcessRequest`.
If yes, it's a security issue.

Reply:
Only Invocable methods, i.e. slots and marked as Q_INVOCABLE. I can also mark all API methods by some TAG (as I've done for public).

Reply:
I think it's safer to use a private TAG instead of using slots for private APIs.

Reply:
>I think it's safer to use a private TAG instead of using slots for private APIs.

I don't understand what you mean.
I still need to call it via invokeMethod so it should be either slot or Q_INVOKABLE.

Reply:
>I don't understand what you mean.

Sorry, I misunderstood something. What I really meant is checking done in `if (!isPublicAPI(methodName) && !sessionActive())` is not enough, you need to also check whether the `methodName` is a valid private API name.

Reply:
Becuase exposing all function in `slots` section for webAPI is error prone IMO, therefore adding a TAG for private API is easier (for us developers) to differentiate which is webAPI and which is normal function.

Reply:
But I still have to be sure that the base class method will not be invoked. I'll think of something.

Reply:
>therefore adding a TAG for private API is easier

But it requires add both PRIVATE/PUBLIC tag and Q_INVOKABLE if we declare it not it slots section:
```c++
WEBAPI_PUBLIC Q_INVOKABLE void login();
```

Reply:
Maybe just have TAGGED methods under slots section? E.g.:
```c++
private slots:
    WEBAPI_PUBLIC void login();
    WEBAPI_PRIVATE void logout();
    WEBAPI_PRIVATE void shutdown();
```

Reply:
>Maybe just have TAGGED methods under slots section? 

Agree.

Reply:
Done.